### PR TITLE
modules: trusted-firmware-a: introduce new module

### PR DESCRIPTION
**Pull request by SgrrZhf**
Add trusted-firmware-a as a module.

This PR is related to https://github.com/zephyrproject-rtos/zephyr/issues/46105

### CODE REVIEW

File: west.yml
Comment (stephanosio) [reply west.yml:231]: Yes, you can temporarily point it to your own repository for testing.

**Review by stephanosio — COMMENTED** [review]
There needs to be samples and tests that exercise this feature.

**Comment by microbuilder** [comment]
@SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

The Corstone-1000 ecosystem FVP is configured with:

- An M0+ bootprocessor ([code available in TF-M](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/platform/ext/target/arm/corstone1000))
- N x Cortex-A35 cores
- A secondary Cortex-M3 based on the [SSE-050 Subsystem](https://developer.arm.com/documentation/100918/0001/introduction/features-of-the-sse-050-subsystem)

We're (Linaro) currently exploring enabling the M3 on this board in Zephyr, using TF-M for the M0+ boot, and eventually we'd like to enable the A35 so that we have an option for A+M on Zephyr communicating with each other, so we can likely help with some of the technical burden here. @povergoing has my contact details if you want to discuss how we can avoid duplicating development efforts on this, if you agree Corstone-1000 FVP might be a good choice.

File: west.yml
Comment (microbuilder) [reply west.yml:231]: You could edit the `west.yml` file to point to your own repo, marking this as DNM while this is the case, with some variation of the instructions below, changing `revision: main` to an appropriate commit hash or branch:

```
For projects that have been setup using ``west``, you can add a local copy of
this module by adding the following sections to ``zephyr/west.yml``:

1. In the ``manifest/remotes`` section add:

.. code-block::

   remotes:
     - name: microbuilder
       url-base: https://github.com/microbuilder

2. In the ``manifest/projects`` section add:

.. code-block::

   - name: linaro_step
     remote: microbuilder
     path: modules/lib/step
     revision: main
```

This should at least let CI point to your repo until there is consensus to fork TF-A into Zephyr's github org.

**Comment by SgrrZhf** [comment]
> There needs to be samples and tests that exercise this feature.

Got that

**Comment by povergoing** [comment]
> @SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

@microbuilder Great idea, no problem, Corstone-1000 is a very powerful platform. I think @SgrrZhf would like to create an issue to plan the future work, including enabling v8A SMP and sure A35 with TF-A and so on. We will contact you if there is some issue:)

**Comment by povergoing** [comment]
> @SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

Free to connect us if you need any help.
BTW, we also need the @stephanosio's help to make it into CI.

File: west.yml
Comment (SgrrZhf) [reply west.yml:231]: Thank you. I'll improve these.

**Comment by stephanosio** [comment]
> BTW, we also need the @stephanosio's help to make it into CI.

I can look into that as part of https://github.com/zephyrproject-rtos/zephyr/pull/45099.

**Comment by SgrrZhf** [comment]
> There needs to be samples and tests that exercise this feature.

Could I create a directory 'samples/tfa-integration' and add a sample named 'hello_world'.  I will make more PR to enable more features such as PSCI later.

In this sample, I enable TFA. This sample will print messages like this and demonstrate that TF-A + zephyr is ok. 

```bash
huizha01@5d2d7be9157a:/workdir/zephyr/build-docker$ west build -t run
-- west build: running target run
[0/1] FVP_Base_RevC-2xAEMvA:
terminal_0: Listening for serial connection on port 5000
terminal_1: Listening for serial connection on port 5001
terminal_2: Listening for serial connection on port 5002
terminal_3: Listening for serial connection on port 5003

Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.flashloader0: FlashLoader: Loaded 168 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/fip.bin'

Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.secureflashloader: FlashLoader: Loaded 25 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/bl1.bin'
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL1: Built : 04:10:00, Jun  7 2022
NOTICE:  BL1: Booting BL2
NOTICE:  BL2: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL2: Built : 04:10:02, Jun  7 2022
NOTICE:  BL1: Booting BL31
NOTICE:  BL31: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL31: Built : 04:10:05, Jun  7 2022
*** Booting Zephyr OS build zephyr-v3.1.0-167-ge78f1ec86872  ***
Hello World! fvp_base_revc_2xaemv8a
```

**Comment by SgrrZhf** [comment]
> > There needs to be samples and tests that exercise this feature.
> 
> Could I create a directory 'samples/tfa-integration' and add a sample named 'hello_world'. I will make more PR to enable more features such as PSCI later.
> 
> In this sample, I enable TFA. This sample will print messages like this and demonstrate that TF-A + zephyr is ok.
> 
> ```shell
> huizha01@5d2d7be9157a:/workdir/zephyr/build-docker$ west build -t run
> -- west build: running target run
> [0/1] FVP_Base_RevC-2xAEMvA:
> terminal_0: Listening for serial connection on port 5000
> terminal_1: Listening for serial connection on port 5001
> terminal_2: Listening for serial connection on port 5002
> terminal_3: Listening for serial connection on port 5003
> 
> Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.flashloader0: FlashLoader: Loaded 168 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/fip.bin'
> 
> Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.secureflashloader: FlashLoader: Loaded 25 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/bl1.bin'
> NOTICE:  Booting Trusted Firmware
> NOTICE:  BL1: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL1: Built : 04:10:00, Jun  7 2022
> NOTICE:  BL1: Booting BL2
> NOTICE:  BL2: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL2: Built : 04:10:02, Jun  7 2022
> NOTICE:  BL1: Booting BL31
> NOTICE:  BL31: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL31: Built : 04:10:05, Jun  7 2022
> *** Booting Zephyr OS build zephyr-v3.1.0-167-ge78f1ec86872  ***
> Hello World! fvp_base_revc_2xaemv8a
> ```

Any suggestions @stephanosio

**Comment by stephanosio** [comment]
> Any suggestions @stephanosio

@SgrrZhf The sample(s) should be more comprehensive than just "hello world." Take a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration for example.

@carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

**Comment by microbuilder** [comment]
> > Any suggestions @stephanosio
> 
> @SgrrZhf The sample(s) should be more comprehensive than just "hello world." Take a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration for example.
> 
> @carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

`tfm_ipc` is probably the simplest example, and you should demonstrate a basic call from NS->S->NS: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration/tfm_ipc

**Comment by carlocaione** [comment]
> @carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

Well, having SMP working (thus PSCI) is also a good prof that TF-A is working correctly.

A long time ago I also added a PSCI test in `./tests/arch/arm64/arm64_psci` that should be good enough IMO.

**Comment by povergoing** [comment]
Agreed with both, but shell we let @SgrrZhf pick only one of two options as the sample for this PR and then enable another one in the next PR? @microbuilder @stephanosio @carlocaione

File: drivers/interrupt_controller/intc_gicv3.c
Comment (SgrrZhf) [inline drivers/interrupt_controller/intc_gicv3.c:491]: CI reports _implicit declaration of function 'sys_read64'_

**Comment by SgrrZhf** [comment]
Hi, @carlocaione @stephanosio  @microbuilder

I'm finished.
Maybe I should put those fix patches to another PR to reduce the current PR's size? 
What do you think about it?

**Comment by carlocaione** [comment]
I'm fine with this but IIUC given that you are modifying the `west.yml` file adding a new remote this cannot be merged as is.

**Comment by SgrrZhf** [comment]
> I'm fine with this but IIUC given that you are modifying the `west.yml` file adding a new remote this cannot be merged as is.

Hi @carlocaione 
I added a temporary remote for testing, and the result is that all tests passed.
I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?

**Comment by carlocaione** [comment]
> Hi @carlocaione I added a temporary remote for testing, and the result is that all tests passed. I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?

@stephanosio @nashif who should fork TF-A? Given that I assume @SgrrZhf is going to maintain it, is that something that he can do himself?

**Comment by SgrrZhf** [comment]
Yes, @povergoing and I volunteer to maintain this(TF-A).

**Comment by povergoing** [comment]
> > Hi @carlocaione I added a temporary remote for testing, and the result is that all tests passed. I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?
> 
> @stephanosio @nashif who should fork TF-A? Given that I assume @SgrrZhf is going to maintain it, is that something that he can do himself?

It would be much better that if there is something we could do it ahead, haha :)

**Comment by stephanosio** [comment]
https://github.com/zephyrproject-rtos/trusted-firmware-a has been created.

**Comment by stephanosio** [comment]
We need to add `West project: trusted-firmware-m` entry in the MAINTAINERS file.

**Comment by SgrrZhf** [comment]
> We need to add `West project: trusted-firmware-a` entry in the MAINTAINERS file.
Got it.

File: boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_smp_ns_defconfig
Comment (povergoing) [inline boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_smp_ns_defconfig:20]: Would you like to remove some CONFIGs that are not needed already, like `CONFIG_UART_PL011`?

File: boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a.dts
Comment (povergoing) [inline boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a.dts:53]: nit: a space before `cpu@300`?

**Review by povergoing — CHANGES_REQUESTED** [review]

File: boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_smp_ns_defconfig
Comment (SgrrZhf) [reply boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_smp_ns_defconfig:20]: Yep

File: boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a.dts
Comment (SgrrZhf) [reply boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a.dts:53]: Ah, I see.

File: MAINTAINERS.yml
Comment (carlocaione) [inline MAINTAINERS.yml:2590]: Can I be added as collaborator? I will be quiet in a corner and I won't bother you, I promise :)

File: soc/arm64/arm/fvp_aemv8a/Kconfig.soc
Comment (carlocaione) [inline soc/arm64/arm/fvp_aemv8a/Kconfig.soc:20]: Is this specific for FVP? If not this should be moved to the core code.

File: tests/arch/arm64/arm64_svc_call/src/main.c
Comment (carlocaione) [inline tests/arch/arm64/arm64_svc_call/src/main.c:24]: nit: shouldn't this be called `smc_call_method_t` ?

File: MAINTAINERS.yml
Comment (povergoing) [reply MAINTAINERS.yml:2590]: Hahaha, do u want to be one of the maintainers?

File: soc/arm64/arm/fvp_aemv8a/Kconfig.soc
Comment (SgrrZhf) [reply soc/arm64/arm/fvp_aemv8a/Kconfig.soc:20]: Right, I'll move them.

File: tests/arch/arm64/arm64_svc_call/src/main.c
Comment (SgrrZhf) [reply tests/arch/arm64/arm64_svc_call/src/main.c:24]: That's right.

File: MAINTAINERS.yml
Comment (stephanosio) [reply MAINTAINERS.yml:2590]: In general, there should be no more than one maintainer.

Comment (carlocaione) [reply MAINTAINERS.yml:2590]: nah, you guys are the experts here, collaborator is fine so that I can be notified of PRs and code changes :) thanks!

Comment (SgrrZhf) [reply MAINTAINERS.yml:2590]: > In general, there should be no more than one maintainer.

I saw there are some area that has more than one maintainers, Could we have two maintainers here?

Comment (stephanosio) [reply MAINTAINERS.yml:2590]: This is not a particularly broad area, so one maintainer would be preferred; but, no strong opinion from me.

Comment (SgrrZhf) [reply MAINTAINERS.yml:2590]: Got it.

**Review by carlocaione — DISMISSED** [review]

**Review by povergoing — DISMISSED** [review]

**Review by microbuilder — DISMISSED** [review]

File: arch/arm64/core/Kconfig
Comment (stephanosio) [inline arch/arm64/core/Kconfig:198]: `default n` is redundant and should not be specified.

Also this config should be prefixed `ARM_` or `ARM64_` since this is specific to this arch.

File: boards/arm64/fvp_base_revc_2xaemv8a/board.cmake
Comment (stephanosio) [inline boards/arm64/fvp_base_revc_2xaemv8a/board.cmake:44]: no new line at the end of file

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: tests/arch/arm64/arm64_smc_call/Kconfig
Comment (stephanosio) [inline tests/arch/arm64/arm64_smc_call/Kconfig:6]: same deal here

File: tests/arch/arm64/arm64_smc_call/testcase.yaml
Comment (stephanosio) [inline tests/arch/arm64/arm64_smc_call/testcase.yaml:11]: These tests are being skipped by every platform (not run at all) because `BUILD_WITH_TFA` is not selected by any platforms.

Comment (stephanosio) [inline tests/arch/arm64/arm64_smc_call/testcase.yaml:11]: These tests are being skipped by every platform (not run at all) because `BUILD_WITH_TFA` is not selected by any platforms.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: arch/arm64/core/Kconfig
Comment (microbuilder) [inline arch/arm64/core/Kconfig:203]: Didn't want to block merging for this, but since there are already some other change requests (thanks @stephanosio  for catching those), this should probably be updated for clarity:

```suggestion
	  VMPIDR_EL2 holds the value of the Virtualization Multiprocessor
	  ID. This is the value returned by EL1 reads of MPIDR_EL1.
	  Normally VMPIDR_EL2 is configured at the EL2 stage.
```

I'm not sure I follow the last line, though, since `SET_VMPIDR_EL2` defaults to `n`,
implying that this isn't 'normally' being set at the EL2 stage. (My experience is more
with Cortex-M, though, so I'm not as familiar with the registers here as I perhaps
should be to comment. The naming and comments aren't helping me, though.)

Comment (microbuilder) [reply arch/arm64/core/Kconfig:203]: @carlocaione Can you perhaps clarify or propose a clearer explanation of what this config setting is intended to do?

Comment (SgrrZhf) [reply arch/arm64/core/Kconfig:203]: That's right, and thanks for clarifying, I'll refine it.

File: tests/arch/arm64/arm64_smc_call/testcase.yaml
Comment (SgrrZhf) [reply tests/arch/arm64/arm64_smc_call/testcase.yaml:11]: Got it.

File: boards/arm64/fvp_base_revc_2xaemv8a/board.cmake
Comment (SgrrZhf) [reply boards/arm64/fvp_base_revc_2xaemv8a/board.cmake]: I didn't see any CMake file that need an extra new line at the end of the file. Do we really need it?

**Review by povergoing — APPROVED** [review]

**Review by carlocaione — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]

**Comment by povergoing** [comment]
Hi @stephanosio, would you like to take another look?

**Review by stephanosio — APPROVED** [review]

**Pull request merged**
